packages/@aws-cdk/aws-appmesh/lib/virtual-gateway.ts

```diff
@@ -0,0 +1,528 @@
+   * mutual exclusivity
+   */
+  public bind(_scope: cdk.Construct): CfnVirtualGateway.VirtualGatewayListenerProperty {
+    return {
+      portMapping: {
+        port: this.port,
+        protocol: Protocol.GRPC,
+      },
+      healthCheck: renderHealthCheck(this.healthCheck, {
+        port: this.port,
+        protocol: Protocol.GRPC,
+      }),
+    };
+  }
+}
+
+/**
+ * Basic configuration properties for a VirtualGateway
+ */
+export interface VirtualGatewayBaseProps {
+  /**
+   * Name of the VirtualGateway
+   *
+   * @default - A name is automatically determined
+   */
+  readonly virtualGatewayName?: string;
+
+  /**
+   * Listeners for the VirtualGateway. Only one is supported.
+   *
+   * @default - Single HTTP listener on port 8080
+   */
+  readonly listeners?: VirtualGatewayListener[];
+
+  /**
+   * Access Logging Configuration for the VirtualGateway
+   *
+   * @default - no access logging
+   */
+  readonly accessLog?: AccessLog;
+}
+
+/**
+ * Properties used when creating a new VirtualGateway
+ */
+export interface VirtualGatewayProps extends VirtualGatewayBaseProps {
+  /**
+   * The mesh which the VirtualGateway belongs to
+   */
+  readonly mesh: IMesh;
+}
+
+/**
+ * Properties required to add multiple GatewayRoutes to a VirtualGateway
+ */
+export interface AddGatewayRouteProps {
+  /**
+   * CloudFormation Logical ID
+   */
+  readonly id: string;
+  /**
+   * Properties to create a GatewayRoute
+   */
+  readonly props: GatewayRouteBaseProps;
+}
+
+abstract class VirtualGatewayBase extends cdk.Resource implements IVirtualGateway {
+  /**
+   * Name of the VirtualGateway
+   */
+  public abstract readonly virtualGatewayName: string;
+
+  /**
+   * The Amazon Resource Name (ARN) for the VirtualGateway
+   */
+  public abstract readonly virtualGatewayArn: string;
+
+  /**
+   * The name of the mesh which the VirtualGateway belongs to
+   */
+  public abstract readonly mesh: IMesh;
+
+  protected readonly listeners = new Array<CfnVirtualGateway.VirtualGatewayListenerProperty>();
+  protected readonly routes = new Array<CfnGatewayRoute>();
+
+  /**
+   * Utility method to add a list of listeners to this VirtualGateway
+   */
+  public addListeners(listeners: VirtualGatewayListener[]) {
+    if (listeners.length + this.listeners.length > 1) {
+      throw new Error('VirtualGateway may have at most one listener');
+    }
+    for (const listener of listeners) {
+      this.addListener(listener);
+    }
+  }
+
+  /**
+   * Utility method to add a single listener to this VirtualGateway
+   */
+  public addListener(listener: VirtualGatewayListener) {
+    if (this.listeners.length > 0) {
+      throw new Error('VirtualGateway may have at most one listener');
+    }
+    this.listeners.push(listener.bind(this));
+  }
+
+  /**
+   * Utility method to add a list of GatewayRoutes to the VirtualGateway
+   */
+  public addGatewayRoutes(props: AddGatewayRouteProps[]): GatewayRoute[] {
```

Comment:
Do we really need this method? Is it worth it to introduce a new type (`AddGatewayRouteProps`) just fo this? Is it a common thing that needs to happen?

Reply:
I thought this was implemented for Virtual Routers, but it was not. I'm going to remove it